### PR TITLE
C# Calculate Task

### PR DESCRIPTION
**Pull request by SanyaKirilv**

### CODE REVIEW

File: CourseApp/CalculateTasks.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+            var listValue = new List<double>();
+            for (double x = StartValue; x <= EndValue; x += DeltaValue)
+            {
+                listValue.Add(x);
+            }
+
+            return listValue;
+        }
+
+        private string StartCalculate(List<double> listValues)
```
Comment (jskonst) [inline CourseApp/CalculateTasks.cs:56]: Please return numeric values, not strings

File: CourseApp/UserInterface.cs [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            Console.WriteLine("Hello!");
+        }
+
+        public bool IsCalculated { get; set; }
+
+        public void StartProgram()
+        {
+            var input = new InputValues();
+            IsCalculated = false;
+            Console.WriteLine("In this program you can:\r\n1 - Calculate with default values\r\n2 - Calculate with custom values\r\n3 - Exit program\r\nWhat you want? Enter integer values.");
```
Comment (alexthvest) [inline CourseApp/UserInterface.cs:20]:
```suggestion            
            Console.WriteLine(string.Join(Environment.NewLine, new string[] {
                  "In this program you can:",
                  "1 - Calculate with default values",
                  "2 - Calculate with custom values",
                  "3 - Exit program",
                  "What you want? Enter integer values."
             });
```

File: CourseApp/CalculateTasks.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+        public double BValue { get; set; }
+
+        public List<(double, double)> StartCalculate(List<double> listValue)
+        {
+            var list = new List<(double, double)>();
+            foreach (double x in listValue)
+            {
+                var value = CalculateValue(x);
+                list.Add((value.Item1, value.Item2));
```
Comment (alexthvest) [inline CourseApp/CalculateTasks.cs:40]:
```suggestion
                list.Add(value);
```

File: CourseApp/UserInterface.cs [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+            }
+        }
+
+        private void Output(List<(double, double)> lst)
+        {
+            Console.WriteLine("Task:");
+            for (int i = 0; i < lst.Count; i++)
+            {
+                Console.WriteLine($"  x = {lst[i].Item1:f2} y = {lst[i].Item2:f2}");
+            }
```
Comment (alexthvest) [inline CourseApp/UserInterface.cs:151]: Use `foreach` and tuple deconstruction (e.g `foreach (var (x, y) in items)`)

File: CourseApp/CheckValues.cs [hunk 5]
```diff
@@ -0,0 +1,55 @@
+namespace CourseApp
+{
+    using System;
+
+    public class CheckValues
+    {
+        public bool IsValidString(string item)
+        {
+            bool isValidValue = false;
```
Comment (jskonst) [inline CourseApp/CheckValues.cs:9]:
```suggestion
            bool isValidValue = true;
```

File: CourseApp/CheckValues.cs [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+            if (string.IsNullOrEmpty(item) || string.IsNullOrWhiteSpace(item))
+            {
+                isValidValue = false;
+                Console.WriteLine("Value is null or empty or contains invalid characters or whitespace. Please enter correct value!");
+            }
+            else
+            {
+                isValidValue = true;
+            }
```
Comment (jskonst) [inline CourseApp/CheckValues.cs:18]:
```suggestion
```

File: CourseApp/UserInterface.cs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+            Output(task.StartCalculate(listValues));
+        }
+
+        private void CalculateCustomValues(List<double> listValues)
+        {
+            Console.Clear();
+            var input = new InputValues();
+            Console.WriteLine("Enter your values:");
+            Console.Write("a = ");
+            var a = input.InputDouble(-Pow(10, 200), Pow(10, 200));
```
Comment (jskonst) [inline CourseApp/UserInterface.cs:103]: use double.MinValue...

File: CourseApp/CheckValues.cs (hunk 6 above)
Comment (SanyaKirilv) [reply CourseApp/CheckValues.cs]: Fixed

File: CourseApp/UserInterface.cs (hunk 7 above)
Comment (SanyaKirilv) [reply CourseApp/UserInterface.cs]: Fixed

File: CourseApp/CalculateTasks.cs (hunk 3 above)
Comment (SanyaKirilv) [reply CourseApp/CalculateTasks.cs]: Fixed

File: CourseApp/CheckValues.cs (hunk 5 above)
Comment (SanyaKirilv) [reply CourseApp/CheckValues.cs]: Fixed

File: CourseApp.Tests/UnitTests.cs [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+        [InlineData(3, 2, 0.36, 1.488)]
+        [InlineData(3, 2, 0.08, 1.565)]
+        [InlineData(3, 2, 0.026, 1.57)]
+        [InlineData(3, 2, 0.35, 1.491)]
+        [InlineData(3, 2, 0.41, 1.471)]
+        [InlineData(3, 2, 0.51, 1.441)]
+        public void TestCalculate(double a, double b, double x, double expected)
+        {
+            var task = new CalculateTasks(a, b);
+            var actual = task.CalculateValue(x).Item2;
```
Comment (alexthvest) [inline CourseApp.Tests/UnitTests.cs:20]:
```suggestion
            var (_, actual) = task.CalculateValue(x);
```

File: CourseApp/Program/Input/CheckValues.cs [hunk 9]
```diff
@@ -0,0 +1,43 @@
+namespace CourseApp.Program.Input
+{
+    using System;
+
+    public class CheckValues
```
Comment (alexthvest) [inline CourseApp/Program/Input/CheckValues.cs:5]: Looks like static class

File: CourseApp.Tests/UnitTests.cs (hunk 8 above)
Comment (SanyaKirilv) [reply CourseApp.Tests/UnitTests.cs:20]: OK

**Review by jskonst — APPROVED** [review]

**Pull request merged**